### DISCUSSION TITLE
Create MeterProvider and TracerProvider separately

### DISCUSSION THREAD

**Discussion by ghost** [open-telemetry/opentelemetry-java#3173] [Q&A]
Hi,

I would like to be able to [specify different resource attributes (`otel.resource.attributes`) for Meter and for Tracer](https://github.bus.zalan.do/SRE/instrumentation-initiative/issues/38). 
In version 0.14.1, we use to use
```
IntervalMetricReader.builder()
                    .readSystemProperties()
                    .setMetricExporter(OtlpGrpcMetricExporter.builder().readSystemProperties().build())
                    .setMetricProducers(Collections.singleton(SdkMeterProvider.builder().buildAndRegisterGlobal()))
                    .build();
```
and 
```
OpenTelemetrySdk.getGlobalTracerManagement().
addSpanProcessor(BatchSpanProcessor
                    .builder(OtlpGrpcSpanExporter.builder().readSystemProperties().build())
                    .readSystemProperties().build());
```

where we would update system properties before each.

With version 1.1.0, creation of MeterProvider and TracerProvider is only possible with using same resource attributes (to use different ones would imply replicating a significant amount of code).
Could you provide such functionality? A way to specify different Resource for MeterProvider and TracerProvider. Or initialise MeterProvider and TracerProvider independently (this way allowing for system properties to be overwritten)?
If so, I can alter the code to provide it.

Thank you for your help in advance.

**Comment by ghost** [comment]
Hey, I'm not 100% sure I understand your use case, but will ty to help :)  

If you're using `autoconfigure` SDK extension, it's currently not possible to have separate resources for tracing and metrics, as the resource is created once for both. 
You can, however, implement a custom `io.opentelemetry.sdk.autoconfigure.spi.SdkMeterProviderConfigurer` (SPI interface). The the implementation may then set a separate resource instance that will be used for metrics. The attributes may be read from your own custom system properties.

If you're not using `autoconfigure` but creating metrics manually, it's just the matter of setting proper Resource during creation of the `SdkMeterProvider`, eg using `SdkMeterProvider.builder().setResource(resource)`.

**Comment by ghost** [comment]
Hi @kubawach ,

thank you for your quick reply !!
I will give a try and let you know if I face any problems :)
One comment on the first suggestion (using `SdkMeterProviderConfigurer`), this would work currently due to the implementation on `OpenTelemetrySdkAutoConfiguration.configureMeterProvider` applying resource from input parameters first and then applying any configurations from these configurers to the builder. Is it intended to keep this order?

**Reply by ghost** [reply]
As far as I can tell order is not specified (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-configuration.md) so currently it's an implementation choice. However if needed I believe that you could open an issue towards the spec to make this guaranteed (@jkwatson is more knowledgable in this topic so I'll leave the confirmation to him).

**Reply by jkwatson** [reply]
Yes, I don't think there are any guarantees right now. Once we make the auto-configure module stable, then we would be very careful about changing the behavior. It does make sense to me that the custom SPI configurer would get applied last, so it had ultimate control over the final configuration, though, so this would probably stay as-is, unless someone has an extremely compelling argument about why we should change it.

**Reply by ghost** [reply]
Thank you for the explanation :) sounds good to me
